### PR TITLE
Pump rework & minor improvements

### PR DESCRIPTION
**Pull request by danielmarbach**
Addresses https://github.com/Particular/NServiceBus.AmazonSQS/issues/180

### CODE REVIEW

File: src/NServiceBus.AmazonSQS/Configure/TransportInfrastructure.cs [hunk 1]
```diff
@@ -71,6 +71,13 @@ public override TransportSendInfrastructure ConfigureSendInfrastructure()
                 () => Task.FromResult(StartupCheckResult.Success));
         }
 
+        public override Task Stop()
+        {
+            sqsClient.Dispose();
+            s3Client.Dispose();
+            return base.Stop();
+        }
```
Comment (mauroservienti) [inline src/NServiceBus.AmazonSQS/Configure/TransportInfrastructure.cs]: any chance `Stop` is called by Core, I suppose, and constructor initialization is not fully completed?
Or put in another way: what happens if the `TransportInfracture.ctor()` throws? Can `sqsClient` and/or `s3Client` be null and `Stop` still be called?

File: src/NServiceBus.AmazonSQS/MessagePump.cs [hunk 2]
```diff
@@ -59,220 +58,257 @@ public MessagePump(ConnectionConfiguration configuration, IAmazonS3 s3Client, IA
...
+                finally
+                {
+                    concurrentReceiveOperations.Clear();
+                }
+            } // while
+        }
+
+        // ReSharper disable once ParameterTypeCanBeEnumerable.Local
+        // ReSharper disable once SuggestBaseTypeForParameter
+        void ProcessMessages(List<Message> receivedMessages, List<Task> concurrentReceiveOperations, CancellationToken token)
```
Comment (mauroservienti) [inline src/NServiceBus.AmazonSQS/MessagePump.cs]: Can we change `List<Message> receivedMessages` to `IEnumerable<Message> receivedMessages`?

File: src/NServiceBus.AmazonSQS/MessagePump.cs [hunk 3]
```diff
@@ -59,220 +58,257 @@ public MessagePump(ConnectionConfiguration configuration, IAmazonS3 s3Client, IA
...
             {
-                await Task.WhenAll(consumerTasks.ToArray()).ConfigureAwait(false);
+                await maxConcurrencySempahore.WaitAsync(token).ConfigureAwait(false);
             }
             catch (OperationCanceledException)
             {
-                // Silently swallow OperationCanceledException
-                // These are expected to be thrown when _cancellationTokenSource
-                // is Cancel()ed above.
+                // shutting, semaphore doesn't need to be released because it was never acquired
```
Comment (mauroservienti) [inline src/NServiceBus.AmazonSQS/MessagePump.cs]: typo in the comment? "shutting" what?

File: src/NServiceBus.AmazonSQS/MessagePump.cs (hunk 2 above)
Comment (danielmarbach) [reply src/NServiceBus.AmazonSQS/MessagePump.cs]: Nope this would mean allocs

File: src/NServiceBus.AmazonSQS/MessagePump.cs (hunk 3 above)
Comment (danielmarbach) [reply src/NServiceBus.AmazonSQS/MessagePump.cs]: Seems to be a merge issue I will fix

File: src/NServiceBus.AmazonSQS/Configure/TransportInfrastructure.cs (hunk 1 above)
Comment (danielmarbach) [reply src/NServiceBus.AmazonSQS/Configure/TransportInfrastructure.cs]: When the ctor throws this happens during Endpoint.Start() which means you have no endpoint instance you can stop

**Comment by danielmarbach** [comment]
@mauroservienti addressed the comment

**Pull request merged**